constraint/pkg/apis/templates/v1/constrainttemplate_types.go

```diff
@@ -0,0 +1,113 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// ConstraintTemplateSpec defines the desired state of ConstraintTemplate
+type ConstraintTemplateSpec struct {
+	CRD     CRD      `json:"crd,omitempty"`
+	Targets []Target `json:"targets,omitempty"`
+}
+
+type CRD struct {
+	Spec CRDSpec `json:"spec,omitempty"`
+}
+
+type CRDSpec struct {
+	Names      Names       `json:"names,omitempty"`
+	Validation *Validation `json:"validation,omitempty"`
+}
+
+type Names struct {
+	Kind       string   `json:"kind,omitempty"`
+	ShortNames []string `json:"shortNames,omitempty"`
+}
+
+type Validation struct {
+	// +kubebuilder:validation:Schemaless
+	// +kubebuilder:validation:Type=object
+	// +kubebuilder:pruning:PreserveUnknownFields
```

Comment:
We should add another bool field to this object (for all versions) called `legacySchema`.

For `v1beta1` and `v1alpha1` it should default to `true`, for `v1` it should default to `false`.

If `legacySchema` is `true`, then we perform x-preserve injections when creating the versionless object.

Doing things this way means that all versions remain wire-compatible and we have no need to write a conversion webhook.

Reply:
done @maxsmythe